prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -0,0 +1,603 @@
+	}
+
+	if (tx_entry->fi_flags & FI_REMOTE_CQ_DATA) {
+		struct rxr_req_opt_cq_data_hdr *cq_data_hdr;
+
+		base_hdr->flags |= RXR_REQ_OPT_CQ_DATA_HDR;
+		cq_data_hdr = (struct rxr_req_opt_cq_data_hdr *)opt_hdr;
+		cq_data_hdr->cq_data = tx_entry->cq_entry.data;
+		opt_hdr += sizeof(*cq_data_hdr);
+	}
+
+	pkt_entry->addr = tx_entry->addr;
+	pkt_entry->hdr_size = opt_hdr - (char *)pkt_entry->pkt;
+}
+
+void rxr_pkt_read_req_hdr(struct rxr_pkt_entry *pkt_entry,
+			  struct rxr_rx_entry *rx_entry)
+{
+	char *opt_hdr;
+	struct rxr_base_hdr *base_hdr;
+
+	rx_entry->addr = pkt_entry->addr;
+
+	base_hdr = rxr_get_base_hdr(pkt_entry->pkt);
+	opt_hdr = (char *)pkt_entry->pkt + RXR_REQ_HDR_SIZE_LIST[base_hdr->type];
+	if (base_hdr->flags & RXR_REQ_OPT_RAW_ADDR_HDR) {
+		rx_entry->rxr_flags |= RXR_REMOTE_SRC_ADDR;
+		opt_hdr += sizeof(struct rxr_req_opt_raw_addr_hdr);
+	}
+
+	if (base_hdr->flags & RXR_REQ_OPT_CQ_DATA_HDR) {
+		struct rxr_req_opt_cq_data_hdr *cq_data_hdr;
+
+		cq_data_hdr = (struct rxr_req_opt_cq_data_hdr *)opt_hdr;
+		rx_entry->rxr_flags |= RXR_REMOTE_CQ_DATA;
+		rx_entry->cq_entry.flags |= FI_REMOTE_CQ_DATA;
+		rx_entry->cq_entry.data = cq_data_hdr->cq_data;
+		opt_hdr += sizeof(struct rxr_req_opt_cq_data_hdr);
+	}
+
+	pkt_entry->hdr_size = opt_hdr - (char *)pkt_entry->pkt;
+	rx_entry->total_len = rxr_pkt_req_total_len(pkt_entry);
+	rx_entry->msg_id = rxr_pkt_rtm_msg_id(pkt_entry);
+	rx_entry->tag = rxr_pkt_rtm_tag(pkt_entry);
+	rx_entry->cq_entry.tag = rx_entry->tag;
+}
+
+size_t rxr_pkt_req_max_data_size(struct rxr_ep *ep, fi_addr_t addr, int pkt_id)
+{
+	struct rxr_peer *peer;
+
+	peer = rxr_ep_get_peer(ep, addr);
+	assert(peer);
+
+	if (rxr_env.enable_shm_transfer && peer->is_local)
+		return rxr_env.shm_max_medium_size;
+
+	return ep->mtu_size - RXR_REQ_HDR_SIZE_LIST[pkt_id]
+		- sizeof(struct rxr_req_opt_raw_addr_hdr)
+		- sizeof(struct rxr_req_opt_cq_data_hdr);
+}
+
+static
+size_t rxr_pkt_req_copy_data(struct rxr_rx_entry *rx_entry,
+			     struct rxr_pkt_entry *pkt_entry,
+			     char *data, size_t data_size)
+{
+	size_t bytes_copied;
+	int bytes_left;
+	/* rx_entry->cq_entry.len is total recv buffer size.
+	 * rx_entry->total_len is from REQ packet and is total send buffer size.
+	 * if send buffer size < recv buffer size, we adjust value of rx_entry->cq_entry.len.
+	 * if send buffer size > recv buffer size, we have a truncated message.
+	 */
+	if (rx_entry->cq_entry.len > rx_entry->total_len)
+		rx_entry->cq_entry.len = rx_entry->total_len;
+
+	bytes_copied = ofi_copy_to_iov(rx_entry->iov, rx_entry->iov_count,
+				       0, data, data_size);
+
+	if (OFI_UNLIKELY(bytes_copied < data_size)) {
+		/* recv buffer is not big enough to hold req, this must be a truncated message */
+		assert(bytes_copied == rx_entry->cq_entry.len &&
+		       rx_entry->cq_entry.len < rx_entry->total_len);
+		rx_entry->bytes_done = bytes_copied;
+		bytes_left = 0;
+	} else {
+		assert(bytes_copied == data_size);
+		rx_entry->bytes_done = data_size;
```

Comment:
Thought this was a copy for the data segment, but this is only used for the initial req packet. Not sure if this affects medium message but will double check when I review that.

Reply:
Yes, = here is correct because this function is used for eager or long message. In either case, the data in req packet is the initial part so = is correct. It will be changed by medium message protocol.